Source/WebCore/dom/FullscreenManager.cpp

```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
 
     INFO_LOG(LOGIDENTIFIER);
 
-    // To achieve that aim, remove all the elements from the top document's stack except for the first before
-    // calling webkitExitFullscreen():
-    RefPtr fullscrenElement = topDocument.fullscreenManager().fullscreenElement();
-    topDocument.fullscreenManager().m_fullscreenElementStack = { WTFMove(fullscrenElement) };
+    m_pendingExitFullscreen = true;
 
-    topDocument.fullscreenManager().exitFullscreen();
+    m_document.eventLoop().queueTask(TaskSource::MediaElement, [this, &topDocument, identifier = LOGIDENTIFIER] {
+        if (!topDocument.page()) {
+            INFO_LOG(LOGIDENTIFIER, "Top document has no page.");
+            return;
+        }
+
+        // This triggers finishExitFullscreen with ExitMode::Resize, which fully exits the document.
+        if (auto* fullscreenElement = topDocument.fullscreenManager().fullscreenElement())
+            page()->chrome().client().exitFullScreenForElement(fullscreenElement);
+        else
+            INFO_LOG(LOGIDENTIFIER, "Top document has no fullscreen element");
+    });
 }
 
-void FullscreenManager::exitFullscreen()
+static Vector<RefPtr<Document>> documentsToUnfullscreen(Document& doc)
```

Comment:
I suggest making the return type `Vector<Ref<Document>>`.

Also, please use the word `document` rather than the abbreviation `doc` for the argument. Also, unclear what the document argument is since the function returns documents. Maybe the argument name should be `firstDocument` or `startDocument` or `elementDocument`.